batch/batch/driver/main.py

```diff
@@ -911,6 +914,125 @@ async def cancel_fast_failing_batches(app):
         await _cancel_batch(app, batch['id'])
 
 
+READY_USER_CORES = pc.Gauge('batch_ready_user_cores', 'Batch ready user cores', ['user', 'inst_coll'])
+RUNNING_USER_CORES = pc.Gauge('batch_running_user_cores', 'Batch running user cores', ['user', 'inst_coll'])
+READY_USER_JOBS = pc.Gauge('batch_ready_user_jobs', 'Batch ready user jobs', ['user', 'inst_coll'])
+RUNNING_USER_JOBS = pc.Gauge('batch_running_user_jobs', 'Batch running user jobs', ['user', 'inst_coll'])
+CREATING_USER_JOBS = pc.Gauge('batch_creating_user_jobs', 'Batch creating user jobs', ['user', 'inst_coll'])
+
+FREE_CORES = pc.Summary('batch_free_cores', 'Batch instance free cores', ['inst_coll'])
+UTILIZATION = pc.Summary('batch_utilization', 'Batch utilization rates', ['inst_coll'])
+ACTUAL_COST_PER_HOUR = pc.Summary('batch_actual_cost_per_hour', 'Batch actual cost ($/hr)', ['inst_coll'])
+BILLED_COST_PER_HOUR = pc.Summary('batch_billed_cost_per_hour', 'Batch billed cost ($/hr)', ['inst_coll'])
+INSTANCES = pc.Gauge('batch_instances', 'Batch instances', ['inst_coll', 'state'])
+
+UserInstCollLabels = namedtuple('UserInstCollLabels', ['user', 'inst_coll'])
+InstCollLabels = namedtuple('InstCollLabels', ['inst_coll'])
+InstanceLabels = namedtuple('InstanceLabels', ['inst_coll', 'state'])
+
+
+async def monitor_user_resources(app):
+    db: Database = app['db']
+
+    ready_user_cores = defaultdict(int)
+    running_user_cores = defaultdict(int)
+    ready_user_jobs = defaultdict(int)
+    running_user_jobs = defaultdict(int)
+    creating_user_jobs = defaultdict(int)
+
+    records = db.select_and_fetchall(
+        '''
+SELECT user, inst_coll,
+  CAST(COALESCE(SUM(ready_cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu,
+  CAST(COALESCE(SUM(n_ready_jobs), 0) AS SIGNED) AS n_ready_jobs,
+  CAST(COALESCE(SUM(running_cores_mcpu), 0) AS SIGNED) AS running_cores_mcpu,
+  CAST(COALESCE(SUM(n_running_jobs), 0) AS SIGNED) AS n_running_jobs,
+  CAST(COALESCE(SUM(n_creating_jobs), 0) AS SIGNED) AS n_creating_jobs
+FROM user_inst_coll_resources
+GROUP BY user, inst_coll;
+'''
+    )
+
+    async for record in records:
+        labels = UserInstCollLabels(user=record['user'], inst_coll=record['inst_coll'])
+        ready_user_cores[labels] += record['ready_cores_mcpu'] / 1000
+        running_user_cores[labels] += record['running_cores_mcpu'] / 1000
+        ready_user_jobs[labels] += record['n_ready_jobs']
+        running_user_jobs[labels] += record['n_running_jobs']
+        creating_user_jobs[labels] += record['n_creating_jobs']
+
+    READY_USER_CORES.clear()
+    RUNNING_USER_CORES.clear()
+    READY_USER_CORES.clear()
+    RUNNING_USER_CORES.clear()
+    CREATING_USER_JOBS.clear()
```

Comment:
I wanted to make sure any values that existed before, but no longer exist are no longer in the gauge. For example, if `user1` had 5 running jobs the last iteration, but now they have 0 running jobs, we should make sure `user1` no longer exists in the gauge.

In this case, since we don't delete any users from the database, it will give the same answer to not clear if we set the value to 0. The plot looked nicer when we just deleted the user from the gauge entirely. Otherwise, you get 50 points that are all at 0.

There's a question here about whether we want to track cores/jobs by user or just by instance collection. I thought it would make it easier to see who ran that huge workload at 2 AM since we don't store that information right now. Thoughts?